### DISCUSSION TITLE
Inject value into options array?

### DISCUSSION THREAD

**Discussion by excid3** [josefarias/hotwire_combobox#127] [Ideas]
Using the combobox as a replacement for a text field, it would be helpful if the value was able to be injected in the options if it didn't exist (when using free-text).

For example, we added this to Hatchbox to suggest a few Cron job examples but use free-text to allow you to enter anything.

<img width="250" alt="Screenshot 2024-03-29 at 8 51 01 PM" src="https://github.com/josefarias/hotwire_combobox/assets/67093/e70bc17e-c3c3-49a5-860f-24ea9d8ae6b6">

When editing, we have to inject the custom value into the array. We have to unique, in order to remove duplicates if they chose one of the default values. We also have to unique to remove a blank value for new records.

```erb
    <%= form.combobox :run_at, [form.object.run_at, "@hourly", "@daily", "@weekly", "@monthly", "@every 4h", "0 12 * * *"].compact.uniq, name_when_new: "cron_job[run_at]", required: true %>
```

This ends up feeling clunky and something the library should handle?

Also, `name_when_new` caught us off guard when it didn't wrap with the form object like Rails fields do. `name_when_new :run_at` is what we expected, not `name_when_new: "cron_job[run_at]"`. Seems like this could be automatic if a model is passed to the form builder to match Rails form fields?

**Comment by josefarias** [comment]
Thanks for the writeup @excid3

A couple of tricky things come to mind about these two points (injecting, and wrapping).

### Re: injecting
Seems like the desired behavior is: if the form object has a value for the specified field — and it's not in the options — then we should insert it into the options.

This should be simple enough for synchronous comboboxes — just add the option if it's not there. But where? Presumably they'd be in alphabetical order but is that always the case? And what if you don't want the value injected? Common instinct is to make it a configurable option. But with the current setup we don't need that config at all, and adding it is not free (as you know, mentioning it for completeness).

Then we'd need to also support this for async comboboxes. And in that case it would be the implementing user's responsibility to add the option as you're doing now (except async), which I'd like to keep all kinds of comboboxes as similar as possible and this would be a departure from that. Or, to keep it similar, we'd need to hijack the streamed responses and add the option somewhere in there —  but then the same questions as above would apply.

That's all doable, of course. But we've been able to keep exceptional behavior across comboboxes to a minimum (even if not 0) — I'm very weary of departing from that from a usability, and also a maintainability standpoint.

Given that this injecting behavior is achievable with no changes, even if a bit unintuitive, I'm leaning towards adding this to a cookbook or wiki of sorts instead of supporting it natively in the library.

Maybe you're able to see the path forward more clearly. Open to external input if so.

### Re: wrapping
Yeah I see how that's surprising. The current setup allows for the most flexibility — you can make the field into literally anything! When the library has run into this kind of scenario before, we've been able to make the most intuitive option an overridable default (your previous suggestion about the automatic `:for_id` in async options comes to mind) — this way it's intuitive by default but it's also super flexible and powerful if you want to override the setting.

The thing with `name_when_new` is that there's no sensible default. So we can't make it overridable if you need the flexibility. If it always wraps the name, then you're limited in what you can do. But if it never wraps the name, then you're free to do whatever.

So similarly to injecting, given that either behavior is achievable with no changes, and the alternative currently looks less flexible, I'm leaning towards documenting this better instead of supporting it natively.

**Reply by excid3** [reply]
**Free Text**

One thing I forgot to mention, I wanted to suggest adding an option to enable the free-text feature, not just `name_when_new`'s presence.

Part of why it felt unintuitive was that `name_when_new` triggered this functionality. It seemed like enabling free text should be its own thing and then customizing the field name would be a second option.

```ruby
# uses `:name` as `name_when_new`
form.combobox :name, free_text: true 
```
Of course, this would only work if wrapping was included, but maybe that's a good separation from the overridden value which can be anything:

```ruby
form.combobox :name, free_text: true, name_when_new: "category[name]"
```

**Missing Value**
I think this is probably worth skipping to be honest. While you could add an `include_existing_value` option to inject into the options, it would also need to have a lambda or something to customize the display value probably and then you're down the rabbit hole of extra functionality.

**Reply by josefarias** [reply]
@excid3 I really like that idea about `free_text`. Sorry it took a while to implement. Will be available in the next version: https://github.com/josefarias/hotwire_combobox/pull/163

Closing this for now. Will reopen if things aren't working as you expected.

Thanks!

**Reply by excid3** [reply]
Nice, thanks for the updates! This has quickly become a field we use all over the place now and allows us to drop tom-select in a lot of cases. 🔥

**Reply by josefarias** [reply]
Love to see it!

I did indeed log into my Hatchbox account to play around with the crontab config 😄 let's keep it going! It's only gonna get better.
